### DISCUSSION TITLE
ElementQuery add `subQueryCriteria` attribute to populate the subQuery directly

### DISCUSSION THREAD

**Discussion by Anubarak** [craftcms/cms#13502] [Ideas] [answered]
I very often have to modify the `subQuery` of  `ElementQueries`directly due to performance reasons (sometimes I just need to join a table in the subQuery and not in the main Query).

This is especially unnecessarily tedious in case the Query is for whatever reasons cloned as `Yii2` does not clone events.
```PHP
$query = Entry::find();
$query->on('afterPrepare', function($event){
    $event->query->leftJoin();
    $event->query->andWhere();
});

// somewhere else in the code
// the clone does not have the event anymore -> it's not filtered correctly
$total = (clone $query)->limit(null)->offset(null)->count();
```

In that case I always have to create a Behavior just in order to left join the subquery. 
(A little bit late but thank you Brandon for creating the `ClonefixTrait`, I read your pull request from 4 years ago to clone Yii2 behaviors when the main class is cloned) 

However: It would be much easier to set properties of the subQuery directly.

```PHP
$query = Entry::find()->withSubQueryCriteria([
    'join' => ['INNER JOIN', ....],
    'where' => [
        // ...
    ]
]);
```
an even better way would be able to pass a Query directly
```PHP
$subQuery = (new Query())
    ->addSelect()
    ->leftJoin()
    ->andWhere();

$query = Entry::find()
    ->withSubQuery($subQuery);
```

and then in your `ElementQuery` class
```PHP
public function prepare($builder): Query
{
    // .....
    $this->query = new Query();
    $this->query->withQueries = $this->withQueries;
    $this->subQuery = $this->passedSubQuery? (clone $this->passedSubQuery) : new Query();
    // .... 
```

That way we could change the subQuery without having to rely on events or behaviors.

**Comment by brandonkelly** [comment, marked as answer]
I ran into this when working on a Craft 5 feature. My solution was to add a new generic `EventBehavior` class, which can be configured with event handlers dynamically, rather than having to create a new event class each time you need it. I just backported that to Craft 4.5, so you don’t need to wait for 5.0. (faedb54f0827c615dbb14a52649ddf1ef418862d)

```php
use craft\behaviors\EventBehavior;
use craft\elements\db\ElementQuery;
use craft\elements\Entry;
use craft\events\CancelableEvent;

$query = Entry::find();

$query->attachBehavior('myBehavior', new EventBehavior([
    ElementQuery::EVENT_AFTER_PREPARE => function(CancelableEvent $event, ElementQuery $query) {
        // ...
    },
], true));
```

(I prefer this over making a bunch of modifications to `Query` itself, which could have unintended consequences and generally would feel more hacky/automagic, whereas this approach is more straightforward as to what’s going on.)

**Reply by Anubarak** [reply]
That's a good idea as well.
Thank you 👍

**Reply by brandonkelly** [reply]
Craft 4.5.0 is out with `EventBehavior` :tada:
